src/Entity/Member.php

```diff
@@ -290,13 +296,23 @@ public function getManagingEmails(): Collection {
         return $this->managingEmails;
     }
 
+    public function getCurrentMembershipStatus(): ?MembershipStatus {
+        return $this->currentMembershipStatus;
+    }
+
+    public function setCurrentMembershipStatus(?MembershipStatus $membershipStatus) {
+        $this->currentMembershipStatus = $membershipStatus;
+    }
+
     /** @see UserInterface */
     public function getUsername(): string { return $this->id; }
 
     /** @see UserInterface */
     public function getRoles(): array {
         $roles = $this->roles;
-        $roles[] = 'ROLE_USER';
+        if ($this->getCurrentMembershipStatus() === null || $this->currentMembershipStatus->getAllowedAccess()) {
+            $roles[] = 'ROLE_USER';
```

Comment:
What is the behaviour if the user has 0 roles? Will that not break things?

Reply:
The user always has the Anonymous/Public user role. And in the case of no roles, it will either deny access or add one of the roles to the array (as the roles property of Member is always an array by default)